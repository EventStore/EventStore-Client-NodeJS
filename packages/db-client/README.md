# @eventstore/db-client

[![npm][npm-badge]][npm-badge-url]
[![Github action CI workflow][ci-badge]][ci-badge-url]
[![license][license-badge]][license-badge-url]

This is the package for the NodeJS client for EventStoreDB 20+ and uses gRPC as the communication protocol.

## Installation

```shell script
# Yarn
$ yarn add @eventstore/db-client

# NPM
$ npm install --save @eventstore/db-client
```

## EventStoreDB Server Compatibility

This client is compatible with version `20.6.1` upwards.

Server setup instructions can be found under the installation section of the [Event Store Docs]. Follow the Docker setup for the simplest configuration.

## Example

The following snippet showcases a simple example where we form a connection, then append and read events from the server.

###### Javascript example:

```javascript
const {
  EventStoreDBClient,
  jsonEvent,
  FORWARDS,
  START,
} = require('@eventstore/db-client');

const client = new EventStoreDBClient({
  endpoint: "localhost:2113",
});

async function simpleTest() {
  const streamName = "es_supported_clients";

  const event = jsonEvent({
    type: "grpc-client",
    data: {
      languages: ["typescript", "javascript"],
      runtime: "NodeJS",
    },
  });

  const appendResult = await client.appendToStream(streamName, [event]);

    // read the event
    const events = client.readStream(streamName, {
      fromRevision: START,
      direction: FORWARDS,
      maxCount: 10,
    });

    for await (const { event } of events) {
      console.log('Appended event: ', event);
    }
  } catch (error) {
    console.error('An error occured: ', error);
  } finally {
    await client.dispose();
  }
})();
```

###### Typescript example:

```typescript
import {
  EventStoreDBClient,
  jsonEvent,
  FORWARDS,
  START,
  JSONEventType,
} from '@eventstore/db-client';

const client = new EventStoreDBClient({
  endpoint: 'localhost:2113',
});

interface Reservation {
  reservationId: string;
  movieId: string;
  userId: string;
  seatId: string;
}

type SeatReservedEvent = JSONEventType<
  'seat-reserved',
  {
    reservationId: string;
    movieId: string;
    userId: string;
    seatId: string;
  }
>;

type SeatChangedEvent = JSONEventType<
  'seat-changed',
  {
    reservationId: string;
    newSeatId: string;
  }
>;

type ReservationEvents = SeatReservedEvent | SeatChangedEvent;

async function simpleTest(): Promise<void> {
  const streamName = 'booking-abc123';

  const event = jsonEvent<SeatReservedEvent>({
    type: 'seat-reserved',
    data: {
      reservationId: 'abc123',
      movieId: 'tt0368226',
      userId: 'nm0802995',
      seatId: '4b',
    },
  });

  const appendResult = await client.appendToStream<ReservationEvents>(
    streamName,
    event
  );

  // By reading the events in the stream, we can construct the current state of the booking

  interface Reservation {
    reservationId: string;
    movieId: string;
    userId: string;
    seatId: string;
  }

  const events = client.readStream<ReservationEvents>(streamName, {
    fromRevision: START,
    direction: FORWARDS,
    maxCount: 10,
  });

  const reservation: Partial<Reservation> = {};

  for await (const { event } of events) {
    switch (event.type) {
      case 'seat-reserved': {
        reservation.reservationId = event.data.reservationId;
        reservation.movieId = event.data.movieId;
        reservation.seatId = event.data.seatId;
        reservation.userId = event.data.userId;
        break;
      }
      case 'seat-changed': {
        reservation.seatId = event.data.newSeatId;
        break;
      }
      default: {
        const _exhaustiveCheck: never = event;
        break;
      }
    }
  }
}

// Do something with our reservation
console.log(reservation);
```

## Support

Information on support can be found on our website: [Event Store Support]

## Communities

- [Discuss]
- [Discord (Event Store)][discord-event-store]
- [Discord (ddd-cqrs-es)][discord-ddd-cqrs-es]

## Contributing

Refer to our [Contribution Guidelines]

[event store support]: https://eventstore.com/support/
[discuss]: https://discuss.eventstore.com/
[discord-event-store]: https://discord.gg/Phn9pmCw3t
[discord-ddd-cqrs-es]: https://discord.com/invite/sEZGSHNNbH
[npm-badge]: https://img.shields.io/npm/v/@eventstore/db-client.svg
[npm-badge-url]: https://www.npmjs.com/package/@eventstore/db-client
[ci-badge]: https://github.com/EventStore/EventStore-Client-NodeJS/workflows/CI/badge.svg?branch=master
[ci-badge-url]: https://github.com/EventStore/EventStore-Client-NodeJS/actions
[license-badge]: https://img.shields.io/npm/l/@eventstore/db-client.svg
[license-badge-url]: https://github.com/EventStore/EventStore-Client-NodeJS/blob/master/LICENSE
[contribution guidelines]: https://github.com/EventStore/EventStore-Client-NodeJS/blob/master/CONTRIBUTING.md
[event store docs]: https://developers.eventstore.com/server/v20.10/docs/installation/