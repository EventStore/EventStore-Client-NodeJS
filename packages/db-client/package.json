{
  "name": "@eventstore/db-client",
  "version": "6.2.1",
  "description": "NodeJS EventStoreDB version 20+ and uses gRPC as the communication protocol.",
  "main": "./dist/index.js",
  "types": "./dist/index.d.ts",
  "scripts": {
    "build": "npm-run-all generate build:ts",
    "build:ts": "tsc --build",
    "build:watch": "npm-run-all generate build:ts-watch",
    "build:ts-watch": "tsc --build --watch",
    "generate": "run-s generate:*",
    "generate:sed-append": "shx sed -i \"s/int64 ([A-z_]*) = ([0-9]*) \\[deprecated = true\\];/int64 \\$1 = \\$2 [deprecated = true, jstype = JS_STRING];/g\" ./protos/*",
    "generate:sed-insert": "shx sed -i \"s/int64 ([A-z_]*) = ([0-9]*);/int64 \\$1 = \\$2 [jstype = JS_STRING];/g\" ./protos/*",
    "generate:folder": "shx mkdir -p ./generated",
    "generate:js": "grpc_tools_node_protoc ./protos/*.proto --proto_path=./protos --grpc_out=grpc_js:./generated --js_out=import_style=commonjs,binary:./generated --ts_out=generate_package_definition:./generated",
    "clean": "run-s clean:*",
    "clean:generated": "shx rm -rf ./generated",
    "clean:dist": "shx rm -rf ./dist",
    "clean:buildinfo": "shx rm -rf *.tsbuildinfo"
  },
  "files": [
    "dist/",
    "generated/"
  ],
  "repository": {
    "type": "git",
    "url": "git+https://github.com/EventStore/EventStore-Client-NodeJS.git"
  },
  "keywords": [
    "EventStore",
    "EventSourcing",
    "Client"
  ],
  "author": "Event Store Limited",
  "license": "Apache-2.0",
  "bugs": {
    "url": "https://github.com/EventStore/EventStore-Client-NodeJS/issues"
  },
  "homepage": "https://developers.eventstore.com/clients/grpc/getting-started?codeLanguage=NodeJS",
  "publishConfig": {
    "access": "public"
  },
  "engines": {
    "node": "^12.18.3 || >=14"
  },
  "dependencies": {
    "@grpc/grpc-js": "^1.9.12",
    "@types/debug": "^4.1.12",
    "@types/google-protobuf": "^3.15.12",
    "@types/node": "^16.18.67",
    "debug": "^4.3.2",
    "google-protobuf": "^3.21.2",
    "uuid": "^8.3.2"
  },
  "devDependencies": {
    "@types/uuid": "^8.3.4",
    "grpc-tools": "^1.12.4",
    "grpc_tools_node_protoc_ts": "^5.3.3",
    "npm-run-all": "^4.1.5",
    "shx": "^0.3.4"
  }
}
