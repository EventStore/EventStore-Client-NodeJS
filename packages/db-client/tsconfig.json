{
  "extends": "../../tsconfig.base.json",
  "compilerOptions": {
    "module": "node16",
    "outDir": "./dist",
    "rootDir": "./src"
  },
  "include": [
    "./src/**/*.ts",
    "./generated"
  ],
  "references": [
    {
      "path": "../db-client-bridge"
    }
  ]
}
