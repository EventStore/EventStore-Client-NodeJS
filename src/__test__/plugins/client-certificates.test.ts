/* eslint-disable @typescript-eslint/no-explicit-any */
import { createTestNode, jsonTestEvents } from "@test-utils";
import { AccessDeniedError, EventStoreDBClient } from "@eventstore/db-client";

describe("client certificates", () => {
  const node = createTestNode();

  beforeAll(async () => {
    await node.up();
  });

  afterAll(async () => {
    await node.down();
  });

  describe("show error if only one of the certificate is provided", () => {
    test("using constructor", async () => {
      try {
        new EventStoreDBClient(
          { endpoint: node.uri },
          {
            rootCertificate: node.certs.root,
            certFile: node.certs.users.admin.certFile,
          }
        );
      } catch (error) {
        expect(error).toMatchInlineSnapshot(
          `[Error: Certificate chain must be given with accompanying private key]`
        );
      }
    });

    test("using connection string", async () => {
      try {
        EventStoreDBClient.connectionString`esdb://${node.uri}?tls=true&tlsCAFile=${node.certPath.root}&certFile=${node.certPath.admin.certPath}`;
      } catch (error) {
        expect(error).toMatchInlineSnapshot(
          `[Error: Both certPath and certKeyPath must be provided together]`
        );
      }
    });
  });

  describe("client initialized with only the admin certificate", () => {
    let client: EventStoreDBClient;

    beforeEach(() => {
      client = new EventStoreDBClient(
        { endpoint: node.uri },
        {
          rootCertificate: node.certs.root,
          certFile: node.certs.users.admin.certFile,
          certKeyFile: node.certs.users.admin.certKeyFile,
        }
      );
    });

    test("using default only certificates", async () => {
      const result = await client.appendToStream(
        "using_default_only_certificates",
        jsonTestEvents(2)
      );
      expect(result).toBeDefined();
    });

    test("overridden user credentials takes precedence over invalid user certificate", async () => {
      await expect(
        client.appendToStream(
          "overridden_user_credentials_takes_precedence",
          jsonTestEvents(2),
          {
            credentials: {
              username: "wrong",
              password: "password",
            },
          }
        )
      ).rejects.toThrow(AccessDeniedError);
    });
  });

  test("user credentials takes precedence over the client certificate during initialization", async () => {
    const clientWithCredentials = new EventStoreDBClient(
      { endpoint: node.uri },
      {
        rootCertificate: node.certs.root,
        certFile: node.certs.users.admin.certFile,
        certKeyFile: node.certs.users.admin.certKeyFile,
      },
      {
        username: "wrong",
        password: "password",
      }
    );

    await expect(
      clientWithCredentials.appendToStream(
        "user_credentials_takes_precedence_over_client_certificate",
        jsonTestEvents(2)
      )
    ).rejects.toThrow(AccessDeniedError);
  });

  test("When the client is initialized with invalid certificate, user credentials take precendence if overriden during a call", async () => {
    const clientWithBadCertificate = new EventStoreDBClient(
      { endpoint: node.uri },
      {
        rootCertificate: node.certs.root,
        certFile: node.certs.users.invalid.certFile,
        certKeyFile: node.certs.users.invalid.certKeyFile,
      }
    );

    expect(
      await clientWithBadCertificate.appendToStream(
        "user_credentials_takes_precedence_over_client_certificate",
        jsonTestEvents(2),
        {
          credentials: {
            username: "admin",
            password: "changeit",
          },
        }
      )
    ).toBeDefined();
  });
});
