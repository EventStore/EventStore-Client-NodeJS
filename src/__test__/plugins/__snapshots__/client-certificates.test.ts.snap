// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`client certificates throw error when constructor is initialised with incorrect channel credentials combinations connection string with certFile only 1`] = `[Error: certPath must be given with accompanying certKeyPath]`;

exports[`client certificates throw error when constructor is initialised with incorrect channel credentials combinations connection string with certKeyFile only 1`] = `[Error: certPath must be given with accompanying certKeyPath]`;

exports[`client certificates throw error when constructor is initialised with incorrect channel credentials combinations constructor initialised with certFile only 1`] = `[Error: Certificate chain must be given with accompanying private key]`;

exports[`client certificates throw error when constructor is initialised with incorrect channel credentials combinations constructor initialised with certKeyFile only 1`] = `[Error: Private key must be given with accompanying certificate chain]`;
